lib/staggered_coarse_op.cu

```diff
@@ -207,28 +242,181 @@ namespace quda {
+    // No need to double spin for the staggered operator, will need to double it for the K-D op.
+    ColorSpinorParam UVparam(T.Vectors(location));
+    UVparam.create = QUDA_ZERO_FIELD_CREATE;
+    UVparam.location = location;
+    UVparam.setPrecision(T.Vectors(location).Precision());
+    UVparam.mem_type = Y.MemType(); // allocate temporaries to match coarse-grid link field
+
+    ColorSpinorField *uv = ColorSpinorField::Create(UVparam);
+
+    ColorSpinorField *av = (dirac == QUDA_STAGGEREDKD_DIRAC || dirac == QUDA_ASQTADKD_DIRAC) ? ColorSpinorField::Create(UVparam) : &const_cast<ColorSpinorField&>(T.Vectors(location));
+    
+    GaugeField *Yatomic = &Y;
+    GaugeField *Xatomic = &X;
+
+    if (Y.Precision() < QUDA_SINGLE_PRECISION) {
+      // we need to coarsen into single precision fields (float or int), so we allocate temporaries for this purpose
+      // else we can just coarsen directly into the original fields
+      GaugeFieldParam param(X); // use X since we want scalar geometry
+      param.location = location;
+      param.setPrecision(QUDA_SINGLE_PRECISION, location == QUDA_CUDA_FIELD_LOCATION ? true : false);
+
+      Yatomic = GaugeField::Create(param);
+      Xatomic = GaugeField::Create(param);
+    }
+
+    // Moving along to the build
+
+    const double kappa = -1.; // cancels a minus sign factor for kappa w/in the dslash application
+    const double mu_dummy = 0.; 
+    const double mu_factor_dummy = 0.;
+
+    bool need_bidirectional = false;
+    if (dirac == QUDA_STAGGEREDKD_DIRAC || dirac == QUDA_ASQTADKD_DIRAC) need_bidirectional = true;
+
+    constexpr QudaGaugeFieldOrder xinvOrder = QUDA_MILC_GAUGE_ORDER;
+    if (XinvKD.FieldOrder() != xinvOrder && XinvKD.Bytes()) errorQuda("unsupported field order %d\n", XinvKD.FieldOrder());
+
+    if (Y.Location() == QUDA_CPU_FIELD_LOCATION) {
+
+      constexpr QudaFieldOrder csOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+      constexpr QudaGaugeFieldOrder gOrder = QUDA_QDP_GAUGE_ORDER;
+
+      if (T.Vectors(Y.Location()).FieldOrder() != csOrder)
+        errorQuda("Unsupported field order %d\n", T.Vectors(Y.Location()).FieldOrder());
+      if (g.FieldOrder() != gOrder) errorQuda("Unsupported field order %d\n", g.FieldOrder());
+
+      typedef typename colorspinor::FieldOrderCB<Float,fineSpin,fineColor,coarseColor,csOrder,vFloat> V;
+      typedef typename colorspinor::FieldOrderCB<Float,fineSpin,fineColor,coarseColor,csOrder,vFloat> F; // will need 2x the spin components for the KD op
+      typedef typename gauge::FieldOrder<Float,fineColor,1,gOrder> gFine;
+      typedef typename gauge::FieldOrder<typename mapper<vFloatXinv>::type,xinvColor,xinvSpin,xinvOrder,true,vFloatXinv> xinvFine;
+      typedef typename gauge::FieldOrder<Float,coarseColor*coarseSpin,coarseSpin,gOrder,true,vFloat> gCoarse;
+      typedef typename gauge::FieldOrder<Float,coarseColor*coarseSpin,coarseSpin,gOrder,true,storeType> gCoarseAtomic;
```

Comment:
General style comment, I think we should be preferring `using` over `typedef` since it's easier to read, e.g., delineation between what we're defining to be what, `using` has the advantage of `=` between.

Reply:
Done.

Reply:
I changed it everywhere I added a `typedef`, and for the time being left other instances unchanged.